rasa/server.py

```diff
@@ -852,6 +853,93 @@ async def retrieve_story(request: Request, conversation_id: Text) -> HTTPRespons
                 f"An unexpected error occurred. Error: {e}",
             )
 
+    @app.get("/billable_conversations")
+    @run_in_thread
+    async def billable_conversations(request: Request) -> HTTPResponse:
+        """Get the number of billable conversations."""
+        start_date = request.args.get("start_date")
+        end_date = request.args.get("end_date")
+        try:
+            # TODO Create a separate component for calculating
+            # and sending billable conversations
+            events = await app.ctx.agent.tracker_store.get_events(
+                start_date, end_date, ["bot", "user", "session_started"]
+            )
+            data = pl.DataFrame(events).lazy()
+            data = (
+                data.with_columns(
+                    (pl.col("type_name") == "session_started")
+                    .cumsum()  # give each session a unique id
+                    .over("sender_id")
+                    .alias("session_ids")
+                )
+                .with_columns(  # turns: user -> bot
+                    pl.when(
+                        (pl.col("type_name") == "bot")
+                        & (pl.col("type_name").shift(1) == "user")
+                    )
+                    .then(1)
+                    .otherwise(0)
+                    .sum()
+                    .over(["sender_id", "session_ids"])
+                    .alias("turns")
+                )
+                .with_columns(
+                    (pl.col("turns") / 50)
+                    .ceil()
+                    .over(["sender_id", "session_ids"])
+                    .alias("convo_turns"),  # 1800 seconds = 30 minutes
+                    ((pl.col("timestamp").max() - pl.col("timestamp").min()) / 1800)
+                    .ceil()
+                    .over(["sender_id", "session_ids"])
+                    .alias("convo_duration"),
+                )
+                .with_columns(  # handle special case when only session_started exists
+                    pl.when(pl.col("convo_duration") < 1)
+                    .then(1)
+                    .otherwise(pl.col("convo_duration"))
+                    .alias("convo_duration")
+                )
+            )
+
+            # sum the convo_turns and convo_duration for each session
+            # the -1 is needed to not double count the base case
+            # session < 30 mins and turns < 50
+            query = (
+                data.groupby(["sender_id", "session_ids"])
+                .agg(
+                    (
+                        (pl.col("convo_turns").max() + pl.col("convo_duration")).max()
```

Comment:
i think this might not be quite right: a conversation that has 100 turns during an hour shouldn't be counted as "3" but rather as "2" (i think)